### PR TITLE
Resolve homebrew and homebrew_cask REGEX build issues for validating …

### LINKED ISSUE #1037: Homebrew Cask does not process cask-names containing hyphens '-' properly
##### SUMMARY
Homebrew Cask does not process cask-names containing hyphens '-' properly. When a hyphen '-' is present, the cask will not be installed claiming that the cask is not present.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
homebrew_cask

##### ANSIBLE VERSION
```
ansible 2.10.1
  config file = /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg
  configured module search path = [u'/Users/michaelwasher/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/michaelwasher/Library/Python/2.7/lib/python/site-packages/ansible
  executable location = /Users/michaelwasher/Library/Python/2.7/bin//ansible
  python version = 2.7.16 (default, Jun  5 2020, 22:59:21) [GCC 4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.29.20) (-macos10.15-objc-
```

##### CONFIGURATION

```
DEFAULT_BECOME(/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg) = False
DEFAULT_BECOME_ASK_PASS(/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg) = False
DEFAULT_BECOME_METHOD(/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg) = sudo
DEFAULT_BECOME_USER(/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg) = root
DEFAULT_HOST_LIST(/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg) = [u'/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/inventory']
DEFAULT_KEEP_REMOTE_FILES(env: ANSIBLE_KEEP_REMOTE_FILES) = True
```

##### OS / ENVIRONMENT
Target OS: Mac OSX 10.15.7

##### STEPS TO REPRODUCE
Run the provided playbook below to install a package in the Mac Environment. It appears as though this fails for all packages containing a '-'.
```
---
- name: Install Homebrew
  hosts: localhost
  become: false
  tasks:
    - name: Install Cask Package through Homebrew
      community.general.homebrew_cask:
        name: visual-studio-code
        state: present
        update_homebrew: true
```

##### EXPECTED RESULTS
This will install the `visual-studio-code` package.

##### ACTUAL RESULTS
This does not install the `visual-studio-code` package, instead logging an error for an invalid-cask.

```
michaelwasher@Michaels-MacBook-Pro ansible_dev_conf % ansible-playbook ./tmp.yml  -vvvv                                                             

ansible-playbook 2.10.1
  config file = /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg
  configured module search path = [u'/Users/michaelwasher/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/michaelwasher/Library/Python/2.7/lib/python/site-packages/ansible
  executable location = /Users/michaelwasher/Library/Python/2.7/bin//ansible-playbook
  python version = 2.7.16 (default, Jun  5 2020, 22:59:21) [GCC 4.2.1 Compatible Apple LLVM 11.0.3 (clang-1103.0.29.20) (-macos10.15-objc-
Using /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/inventory as it did not pass its verify_file() method
script declined parsing /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/inventory as it did not pass its verify_file() method
auto declined parsing /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/inventory as it did not pass its verify_file() method
Set default localhost to localhost
Parsed /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/inventory inventory source with ini plugin
Loading collection community.general from /Users/michaelwasher/.ansible/collections/ansible_collections/community/general
Loading callback plugin default of type stdout, v2.0 from /Users/michaelwasher/Library/Python/2.7/lib/python/site-packages/ansible/plugins/callback/default.pyc

PLAYBOOK: tmp.yml **************************************************************************************************************************************************************************************************************************
Positional arguments: ./tmp.yml
become_method: sudo
inventory: (u'/Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/inventory',)
forks: 5
tags: (u'all',)
verbosity: 4
connection: smart
timeout: 10
1 plays in ./tmp.yml

PLAY [Install Homebrew] ********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] *********************************************************************************************************************************************************************************************************************
task path: /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/tmp.yml:2
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: michaelwasher
<localhost> EXEC /bin/sh -c 'echo ~michaelwasher && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/michaelwasher/.ansible/tmp `"&& mkdir "` echo /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770695.22-70144-75935192927992 `" && echo ansible-tmp-1601770695.22-70144-75935192927992="` echo /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770695.22-70144-75935192927992 `" ) && sleep 0'
<localhost> Attempting python interpreter discovery
<localhost> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
<localhost> Python interpreter discovery fallback (unsupported platform for extended discovery: darwin)
Using module file /Users/michaelwasher/Library/Python/2.7/lib/python/site-packages/ansible/modules/setup.py
<localhost> PUT /Users/michaelwasher/.ansible/tmp/ansible-local-70137AWmosn/tmp8Bs5Lu TO /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770695.22-70144-75935192927992/AnsiballZ_setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770695.22-70144-75935192927992/ /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770695.22-70144-75935192927992/AnsiballZ_setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770695.22-70144-75935192927992/AnsiballZ_setup.py && sleep 0'
[WARNING]: Platform darwin on host localhost is using the discovered Python interpreter at /usr/bin/python, but future installation of another Python interpreter could change the meaning of that path. See
https://docs.ansible.com/ansible/2.10/reference_appendices/interpreter_discovery.html for more information.
ok: [localhost]
META: ran handlers

TASK [Install Cask Package through Homebrew] ***********************************************************************************************************************************************************************************************
task path: /Users/michaelwasher/Desktop/UserFolder/ansible_dev_conf/tmp.yml:6
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: michaelwasher
<localhost> EXEC /bin/sh -c 'echo ~michaelwasher && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/michaelwasher/.ansible/tmp `"&& mkdir "` echo /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770696.88-70178-251044235237686 `" && echo ansible-tmp-1601770696.88-70178-251044235237686="` echo /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770696.88-70178-251044235237686 `" ) && sleep 0'
Using module file /Users/michaelwasher/.ansible/collections/ansible_collections/community/general/plugins/modules/homebrew_cask.py
<localhost> PUT /Users/michaelwasher/.ansible/tmp/ansible-local-70137AWmosn/tmpoAdY3T TO /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770696.88-70178-251044235237686/AnsiballZ_homebrew_cask.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770696.88-70178-251044235237686/ /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770696.88-70178-251044235237686/AnsiballZ_homebrew_cask.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /Users/michaelwasher/.ansible/tmp/ansible-tmp-1601770696.88-70178-251044235237686/AnsiballZ_homebrew_cask.py && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "accept_external_apps": false, 
            "greedy": false, 
            "install_options": [], 
            "name": [
                "visual-studio-code"
            ], 
            "path": "/usr/local/bin", 
            "state": "present", 
            "sudo_password": null, 
            "update_homebrew": true, 
            "upgrade_all": false
        }
    }, 
    "msg": "Invalid cask: visual-studio-code."
}

PLAY RECAP *********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0 
```

### PR DESCRIPTION
**Pull request by MichaelWasher**
…cask-names.

##### SUMMARY

Change '-' character used for REGEX build to resolve a bug.

Fixes #1037

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

homebrew_cask
homebrew

### CODE REVIEW

**Review by martinm82 — COMMENTED** [review]
Please add some regression tests. I am also wondering whether a similar problem occurs with packages including dots in the Name (never have seen such a package).

**Comment by felixfontein** [comment]
Please also add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to).

About tests: there are only integration tests for homebrew, and no tests at all for homebrew_cask. For this, probably unit tests would be great. If you want to create basic unit tests for this, we're happy to help (https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html#unit-tests for basic instructions, https://github.com/ansible-collections/community.general/tree/main/tests/unit/plugins/modules/packaging/os is where tests belong; for the test it probably suffices to create instances of the `Homebrew`/`HomebrewCask` classes with `None` passed as everything in the constructor, and then run some tests against the compiled regexes). If you don't want to do this, I fully understand, it's quite a lot to ask for someone who just wants to fix a bug. :)

**Comment by MichaelWasher** [comment]
No it's alright, I'll whip up some tests in the coming days. Just swamped with other work at the moment but will write up some tests in the coming days. I was also thinking the same with `.` but then thought if it's interpreted as wildcard then it would fail all casks.

So I saw this.
~~~ https://docs.python.org/3/library/re.html#search-vs-match
Special characters lose their special meaning inside sets. For example, [(+*)] will match any of the literal characters '(', '+', '*', or ')'
~~~

And tested here:
~~~
>>> INVALID_CASK_REGEX = _create_regex_group(".")
>>> INVALID_CASK_REGEX.search(".")
>>> INVALID_CASK_REGEX.search("t")
<_sre.SRE_Match object at 0x1075e0850>
~~~

I think it's just the `-` that's affected within the grouping brackets `[]`.

**Comment by yinzara** [comment]
I noticed that Ansible 2.9 `homebrew_cask` works fine with the current version of community.general (i.e. 1.2.0) however Ansible 2.10 breaks with the given error message.

Is there a way for us to verify that this PR works for both Ansible 2.9 and 2.10?

**Comment by felixfontein** [comment]
> Is there a way for us to verify that this PR works for both Ansible 2.9 and 2.10?

Add tests ;-) Right now, the homebrew_cask module has no tests, so if nobody tests it manually, there are always risks of something breaking.

**Comment by MichaelWasher** [comment]
@felixfontein ; I'm new to testing in Ansible environment, does this repo integrate into a CI/CD system for automated testing? 
Or just running the `ansible-test units --docker default -v` is sufficient? 
Also does running this command test on both the Ansible 2.9 and 2.10, or is this achieved by running the command with the `ansible-test` command from version?

**Comment by felixfontein** [comment]
@MichaelWasher this repo uses Shippable to run tests. Tests are run with Ansible 2.9, 2.10 and ansible/ansible's `devel` branch.

For unit tests, `ansible-test units --docker default -v` is sufficient. For integration tests, it depends on the target. Not all targets are available as docker containers, like RHEL, FreeBSD, and macOS and OSX. These are VMs that cannot be accessed directly.

**Comment by felixfontein** [comment]
> Also does running this command test on both the Ansible 2.9 and 2.10, or is this achieved by running the command with the ansible-test command from version?

It uses ansible-test from these Ansible versions.

**Comment by felixfontein** [comment]
(Usually it's fine to run unit tests for one of these Ansible versions, even with a restricted list of Python versions, before letting CI run all of them.)

**Comment by felixfontein** [comment]
The tests look good! Could you add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to) as well?

**Comment by felixfontein** [comment]
Sanity tests report (pep8):
```
tests/unit/plugins/modules/packaging/os/test_homebrew_cask.py:9:1: E302: expected 2 blank lines, found 1
```

**Comment by MichaelWasher** [comment]
Will do, thank you for your help with this! Sorry about the delay, l've been a big bogged up with work.

**Comment by felixfontein** [comment]
@MichaelWasher no problem, take your time! The next release will be in ~one month, so you have that much time to complete this ;)

File: changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation [hunk 1]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+- homebrew - fix package name validation for packages and casks containing hypen `-`
+  (https://github.com/ansible-collections/community.general/pull/1038)
```
Comment (felixfontein) [inline changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation:3]:
```suggestion
- homebrew - fix package name validation for packages and casks containing hypen ``-``
  (https://github.com/ansible-collections/community.general/issues/1037).
- homebrew_cask - fix package name validation for packages and casks containing hypen ``-``
  (https://github.com/ansible-collections/community.general/issues/1037).
```

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by felixfontein** [comment]
Is this still a WIP? If not, please remove that from the PR title :)

File: plugins/modules/packaging/os/homebrew.py [hunk 2]
```diff
@@ -168,7 +168,7 @@ class HomebrewException(Exception):
 
 
 # utils ------------------------------------------------------------------- {{{
-def _create_regex_group(s):
+def _create_regex_group_compliment(s):
```
Comment (jmknoble) [inline plugins/modules/packaging/os/homebrew.py:171]: You've misspelled [complement](https://en.wikipedia.org/wiki/Complement_(set_theory)) here.  (*compliment* == "Those shoes look nice on you!")

Comment (felixfontein) [reply plugins/modules/packaging/os/homebrew.py:171]: Well, it's a nice function, so maybe...? ;)

**Comment by MichaelWasher** [comment]
Should be done now . Thanks!!

File: changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation (hunk 1 above)
Comment (felixfontein) [reply changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation]: @MichaelWasher this is still open.

**Comment by felixfontein** [comment]
Please see my "outdated" comment on the changelog fragment from yesterday.

File: changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation (hunk 1 above)
Comment (MichaelWasher) [reply changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation]: I don't understand what is being asked sorry. You would prefer to 4 lines instance?

Comment (MichaelWasher) [reply changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation]: I changed it to the 4 lines changelog just incase
https://github.com/ansible-collections/community.general/pull/1038/files#diff-961d0154e27fe6fd59ec4e79ec36fb5de7a0fa2689a5e8f92288fcbde3af2ee4R1-R5

Comment (felixfontein) [reply changelogs/fragments/1038-fix-homebrew-and-homebrew-cask-package-validation]: I'm also fine if you put it in two lines :) The problems with your changelog fragment were:
1. You only mentioned the homebrew module, not the homebrew_cask module.
2. You didn't use correct formatting for the dash (this is reStructuredText, not MarkDown).
3. A dot is missing at the end of the entry.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@martinm82 @yinzara any more comments?

**Review by martinm82 — APPROVED** [review]

**Comment by martinm82** [comment]
LGTM

**Review by jschell12 — APPROVED** [review]

**Comment by felixfontein** [comment]
@MichaelWasher thanks for fixing this!
@martinm82 @yinzara @jmknoble @jschell12 thanks for reviewing and testing!

**Comment by danielholdsworth** [comment]
Forgive my ignorance guys but how are we to install the community module with the fix? it's not appearing in ansible-galaxy yet is all.

**Comment by jschell12** [comment]
This is how I’m doing it in requirements.yaml:
``` collections:
  - name: https://github.com/ansible-collections/community.general
    type: git
    version: 4c379bd3b2147710c3e7a1d900cddf47a6b7cd42 
   ```

You should be able to reference 1.3.0 by end of November

**Comment by danielholdsworth** [comment]
@jschell12 Ok cheers for that, I've ended up just creating a tar of the current state of the repo, and installing that as a local collection.

That way I can freeze this working version of the collection (we only really use the homebrew modules) and had it over to other devs etc easily enough.

**Pull request merged**